service/environment.go

```diff
@@ -141,7 +145,35 @@ func (e *environment) DefaultScheme() string {
 }
 
 func (e *environment) Instancers() Instancers {
-	return e.instancers.Copy()
+	e.lock.RLock()
+	instancersCopy := e.instancers.Copy()
+	e.lock.RUnlock()
+	return instancersCopy
+}
+
+func (e *environment) UpdateInstancers(currentKeys map[string]bool, instancersToAdd Instancers) {
+	// add new instancers
+
+	for key, value := range instancersToAdd {
+		e.lock.Lock()
+		e.instancers.Set(key, value)
+		e.lock.Unlock()
+	}
+
+	// remove outdated instancers
+	for key := range e.instancers {
```

Comment:
shouldn't we be in at least a read lock in order to run through all the instancers? we're not looking at the values, but technically the keys could be changing, right?

Reply:
The set of instancers, along with most other Environment state, should be immutable after configuration is read.

Reply:
Actually now I see what's going on with `UpdateInstancers`.  Yeah, you're right, a `sync.RWMutex` should probably be used here to guard the set of instancers.